[CORE-12307] `rptest`: deflake `compaction_recovery_test`

WillemKauf opened the pull request:
Per the class comment, we want to disable adjacent segment compaction to prevent indices from constantly being rebuilt/concatenated in this test.

However, this value was only being set after a restart, which means the initial `wait_until()` dependent on the production of `segment`s could flake if `segment`s were constantly being adjacently merged together (reducing the `segment` count below what is being awaited).

Set this value from the start of the test to disable adjacent segment merging for its entirety.

## Backports Required

- [X] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.2.x
- [ ] v25.1.x
- [ ] v24.3.x
- [ ] v24.2.x

## Release Notes

* none

WillemKauf committed:
`rptest`: better logging in `compaction_recovery_test`

WillemKauf commented:
/ci-repeat 5
skip-units
tests/rptest/tests/compaction_recovery_test.py::CompactionRecoveryTest.test_index_recovery
dt-repeat=20

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--ad5d76f8-1b7b-474a-b732-bb919e8b775d-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/69669">build#69669</a></summary>


| test_class             | test_method         | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:-----------------------|:--------------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| CompactionRecoveryTest | test_index_recovery | null             | integration | https://buildkite.com/redpanda/redpanda/builds/69669#0198420d-7cc1-48d8-98ac-f3a60ed8265c | FLAKY         | 39/40    | upstream reliability is '99.44903581267218'. current run reliability is '95.23809523809523'. drift is 4.21094 and the allowed drift is set to 50. The test should PASS |
| CompactionRecoveryTest | test_index_recovery | null             | integration | https://buildkite.com/redpanda/redpanda/builds/69669#01984213-8eb8-48c8-960f-317d817e6280 | FLAKY         | 38/40    | upstream reliability is '99.44903581267218'. current run reliability is '90.47619047619048'. drift is 8.97285 and the allowed drift is set to 50. The test should PASS |
| CompactionRecoveryTest | test_index_recovery | null             | integration | https://buildkite.com/redpanda/redpanda/builds/69669#01984213-8eb9-4262-881e-75e9e2c54894 | FLAKY         | 38/40    | upstream reliability is '99.44903581267218'. current run reliability is '90.47619047619048'. drift is 8.97285 and the allowed drift is set to 50. The test should PASS |
</details>
<!--/body-->

WillemKauf committed:
`rptest`: set `max_compacted_log_segment_size=1` in `compaction_recovery_test`

Per the class comment above, we want to disable adjacent segment compaction
to prevent indices from constantly being rebuilt/concatenated in this test.

However, this value was only being set after a restart, which means the initial
`wait_until()` dependent on the production of `segment`s could flake if
`segment`s were constantly being adjacently merged together (reducing the
`segment` count below what is being awaited).

Set this value from the start of the test to disable adjacent segment merging
for its entirety.

WillemKauf commented:
/ci-repeat 5
skip-units
tests/rptest/tests/compaction_recovery_test.py::CompactionRecoveryTest.test_index_recovery
dt-repeat=20

WillemKauf commented:
No flakes in Build 69677

oleiman reviewed (approved):
nice one 👍
